File: src/my_project/interviews/top_150_questions_round_22/ex_138_minimum_path_sum.py
from typing import List

class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        dp = [row[:] for row in grid]
        for i in range(1, m):
            dp[i][0] += dp[i - 1][0]
        for j in range(1, n):
            dp[0][j] += dp[0][j - 1]
        for i in range(1, m):
            for j in range(1, n):
                dp[i][j] += min(dp[i - 1][j], dp[i][j - 1])
        return dp[m - 1][n - 1]

File: src/my_project/interviews_typescript/top_150_questions_round_1/ex_138_minimum_path_sum.ts
function minPathSum(grid: number[][]): number {
    const m = grid.length, n = grid[0].length;
    const dp: number[][] = grid.map(row => [...row]);
    for (let i = 1; i < m; i++) dp[i][0] += dp[i - 1][0];
    for (let j = 1; j < n; j++) dp[0][j] += dp[0][j - 1];
    for (let i = 1; i < m; i++) {
        for (let j = 1; j < n; j++) {
            dp[i][j] += Math.min(dp[i - 1][j], dp[i][j - 1]);
        }
    }
    return dp[m - 1][n - 1];
};